### PR TITLE
scm: select node based on active editor

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixes: https://github.com/eclipse-theia/theia/issues/11558.

The pull-request updates the `scm-view` to properly select the appropriate `node` in the changes tree based on the active editor. 

#### How to test

1. start the application with a workspace under git version control
2. perform changes in two files
3. open the `scm-view` and open the `working tree` changes for both
4. switching between the two files should now properly set the tree selection
5. confirm it works for `list` and `tree` view modes
6. confirm it works for deep nested files
7. confirm it works for working tree, regular editors, and preview editors

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

File: packages/scm/src/browser/scm-widget.tsx [hunk 1]
```diff
@@ -84,6 +86,14 @@ export class ScmWidget extends BaseWidget implements StatefulWidget {
                 this.updateViewMode(e.newValue);
             }
         }));
+        this.toDispose.push(this.shell.onDidChangeCurrentWidget(({ newValue }) => {
+            if (newValue instanceof EditorWidget) {
+                const uri = newValue.getResourceUri();
+                if (uri) {
+                    this.resourceWidget.selectNodeByUri(uri);
+                }
+            }
```
Comment (msujew) [inline packages/scm/src/browser/scm-widget.tsx:95]: Suggestion: I believe we can simplify this (and make it less specific to `EditorWidget`).

```suggestion
            const uri = NavigatableWidget.getUri(newValue || undefined);
            if (uri) {
                this.resourceWidget.selectNodeByUri(uri);
            }
```

File: packages/scm/src/browser/scm-tree-widget.tsx [hunk 2]
```diff
@@ -355,6 +358,19 @@ export class ScmTreeWidget extends TreeWidget {
...
+    selectNodeByUri(uri: URI): void {
+        const root = this.model.root;
+        if (!root) { return; }
+        for (const node of new DepthFirstTreeIterator(root)) {
+            if (ScmFileChangeNode.is(node)) {
+                if (node.sourceUri.includes(uri.path.toString())) {
+                    this.model.selectNode(node);
+                    return;
+                }
+            }
```
Comment (msujew) [inline packages/scm/src/browser/scm-tree-widget.tsx:370]: Minor:
```suggestion
            if (ScmFileChangeNode.is(node) && node.sourceUri.includes(uri.path.toString())) {
                this.model.selectNode(node);
                return;
            }
```

**Comment by vince-fugnitto** [comment]
@msujew thank you for the review, I incorporated your suggestions and confirmed it works.

**Review by msujew — APPROVED** [review]
Looks good to me 👍

**Review by colin-grant-work — APPROVED** [review]
Based on our offline discussions, this appears to be the best way to achieve this end. 👍

**Pull request merged**